proposals/0000-exception-backtraces.rst

```diff
@@ -0,0 +1,678 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Export the following new definitions from ``Control.Exception`` which provide a
+convenient way to gain access to ``ExceptionContext`` in exception handlers: ::
+
+  data ExceptionWithContext a =
+    ExceptionWithContext ExceptionContext a
+
+  instance Show a => Show (ExceptionWithContext a)
+
+  instance Exception a => Exception (ExceptionWithContext a) where
+      toException (ExceptionWithContext ctxt e) = SomeException e
+        where ?exceptionContext = ctxt
+      fromException se = do
+          e <- fromException se
+          return (ExceptionWithContext (exceptionContext se) e)
+      displayException = displayException . toException
+
+Preserving context on rethrowing
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In ``Control.Exception``, modify existing definitions as follows:
+
+* Modify ``catch`` to preserve ``ExceptionContext`` when a handler rethrows an
+  exception: ::
+
+    catch :: Exception e => IO a -> (e -> IO a) -> IO a
+    catch (IO io) handler = IO $ catch# io handler'
+     where
+       handler' e =
+         case fromException e of
+           Just e' -> unIO (annotateIO (someExceptionContext e) (handler e'))
+           Nothing -> raiseIO# e
+
+  Modify ``catchJust`` and ``handleJust`` accordingly (mutatis mutandis).
+
+Capturing Backtraces on Exceptions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In ``Control.Exception``, modify existing definitions as follows:
+
+* Add the following method and default definition to the ``Exception``
+  typeclass: ::
+
+    backtraceDesired :: e -> Bool
+    backtraceDesired _ = True
+
+* Modify ``throwIO`` as follows: ::
+
+    -- This type will be further refined in section 2.8
+    throwIO :: forall e a. Exception e => e -> IO a
+    throwIO e = do
+        ?exceptionContext <-
+          if backtraceDesired e
+            then do
+              bt <- collectBacktraces
+              return (addExceptionAnnotation bt emptyExceptionContext)
+            else return emptyExceptionContext
+        let se :: SomeException
+            se = SomeException e
+        raiseIO# se
+
+Export the following new definitions from ``Control.Exception``:
+
+* The following ``newtype`` wrapper and instance which can be used by the user
+  when throwing an exception to disable backtrace collection: ::
+
+    newtype NoBacktrace e = NoBacktrace e
+
+    instance Show e => Show (NoBacktrace e)
+
+    instance Exception e => Exception (NoBacktrace e) where
+      fromException = NoBacktrace . fromException
+      toException (NoBacktrace e) = toException e
+      backtraceDesired _ = False
+
+``HasCallStack`` Backtraces for Thrown Exceptions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In ``Control.Exception`` add ``HasCallStack`` constraints to the exception
+``throw`` functions to allow inclusion in backtrace context: ::
+
+    throwIO :: forall e a. (HasCallStack, Exception e) => e -> IO a
+    throw   :: forall e a. (HasCallStack, Exception e) => e -> a
```

Comment:
I assume that you intend `throw` to collect backtraces like `throwIO` does, but this is not specified.

How about `error`?

Reply:
A very good point since `error` is not implemented in terms of `throw`.

This also raises the question of what to do about the `ErrorCallWithLocation` constructor, which carries a backtrace. In the interest of keeping the scope of the proposal at bey, I propose to leave it as-is currently. This means that there will be two backtraces produced by `error`: one in the `ExceptionContext` and another in the `ErrorCallWithLocation` constructor.

Reply:
A very good point since `error` is not implemented in terms of `throw`.

This also raises the question of what to do about the `ErrorCallWithLocation` constructor, which carries a backtrace. In the interest of keeping the scope of the proposal at bey, I propose to leave it as-is currently. This means that there will be two backtraces produced by `error`: one in the `ExceptionContext` and another in the `ErrorCallWithLocation` constructor.

Reply:
This also raises the question of what should happen with `errorWithoutStackTrace`. I will propose that we keep it unchanged for now, meaning that it will not carry a `Backtrace` in its `ExceptionContext`.

Reply:
I think both are quite fair. At any rate, `error` is not typically called in a performance sensitive situation as it's more of a crash and burn type of thing.